internal/pkg/agent/cmd/unprivileged.go

```diff
@@ -0,0 +1,127 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package cmd
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"runtime"
+
+	"github.com/spf13/cobra"
+
+	"github.com/elastic/elastic-agent/internal/pkg/agent/application/paths"
+	"github.com/elastic/elastic-agent/internal/pkg/agent/install"
+	"github.com/elastic/elastic-agent/internal/pkg/cli"
+	"github.com/elastic/elastic-agent/pkg/control/v2/client/wait"
+	"github.com/elastic/elastic-agent/pkg/utils"
+)
+
+const (
+	endpoint = "endpoint"
+)
+
+func newUnprivilegedCommandWithArgs(s []string, streams *cli.IOStreams) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "unprivileged",
+		Short: "Switch installed Elastic Agent to run as unprivileged",
+		Long: `This command converts the installed Elastic Agent from running privileged to running as unprivileged.
+
+By default this command will ask or a confirmation before making this change. You can bypass the confirmation request
+using the -f flag. This will always stop the running Elastic Agent (if running) before performing the switch it is
+possible that loss of metrics could occur during this small window of time. The command also always starts the
+Elastic Agent at the end (even if it was off to start). In the case that the Elastic Agent is already running
+unprivileged it will still perform all the same work, including stopping and starting the Elastic Agent.
+`,
+		Args: cobra.ExactArgs(0),
+		Run: func(c *cobra.Command, args []string) {
+			if err := unprivilegedCmd(streams, c); err != nil {
+				fmt.Fprintf(streams.Err, "Error: %v\n%s\n", err, troubleshootMessage())
+				os.Exit(1)
+			}
+		},
+	}
+
+	cmd.Flags().BoolP("force", "f", false, "Do not prompt for confirmation")
+	cmd.Flags().DurationP("daemon-timeout", "", 0, "Timeout waiting for Elastic Agent daemon")
+
+	return cmd
+}
+
+func unprivilegedCmd(streams *cli.IOStreams, cmd *cobra.Command) (err error) {
+	isAdmin, err := utils.HasRoot()
+	if err != nil {
+		return fmt.Errorf("unable to perform unprivileged command while checking for root/Administrator rights: %w", err)
+	}
+	if !isAdmin {
+		return fmt.Errorf("unable to perform unprivileged command, not executed with %s permissions", utils.PermissionUser)
+	}
+
+	// TODO(blakerouse): More work to get this working on macOS.
+	// Need to switch the vault from keystore based to file based vault.
+	if runtime.GOOS == "darwin" {
+		return errors.New("unable to perform unprivileged on macOS (not supported)")
+	}
+
+	// cannot switch to unprivileged when Elastic Defend exists in the policy
+	err = ensureNoElasticDefend()
```

Comment:
👍

Reply:
Normally osquery is running as privileged. Quite a lot of functionality will be unavailable if running unprivileged or returning different results. 

Osqueryd also checks for "safe permissions" when it loads our custom extension.

```
 When osquery launches, each path is evaluated for "safe permissions" 
(extension executable files must be owned by root or Administrator) 
and executed as a monitored child process
```
https://osquery.readthedocs.io/en/stable/deployment/extensions/#autoloading-extensions

```
If you observe a runtime error from osquery, Extension binary has unsafe permissions, 
you have to lock down the filesystem permissions on the extension executable. 
See the steps in "Extensions Binary Permissions," above.
```
https://osquery.readthedocs.io/en/stable/deployment/extensions/#troubleshooting

Reply:
For osquery, you should mark the integration as requiring root privileges like defend does in https://github.com/elastic/endpoint-package/blob/ce921805467ab55953adc3721e7db8581915bd35/package/endpoint/manifest.yml#L23-L25. You can also do this for individual data streams but I don't think that applies to OSquery.

You could also add a prevention to the spec file to prevent us from even starting osquery if you aren't root, but since it doesn't fundamentally not work when not root, this might not be necessary.

Reply:
If you do update the spec, you would want to update the agentbeat spec.